### PR TITLE
support adam bf16 state

### PR DESCRIPTION
**Pull request by XiaobingSuper**
# Description

This PR supports Adam bf16 state to reduce memory usage, we have tested some smaller models or largest models (LLM models), and we get a similar convergence even with this bf16 lower precision state. For the deepseek-v3 report, they also used the bf16 state to reduce training memory usage.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Change A
- Change B

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by BestJuly** [comment]
Hi @XiaobingSuper , is there any updates on this PR? Saving GPU memories using BF16 is valuable. If you are running out of time, I am glad to help make some changes according to Tim's suggestions based on your branch, to see whether this PR can be merged ASAP. Thank you.

**Comment by XiaobingSuper** [comment]
> Hi @XiaobingSuper , is there any updates on this PR? Saving GPU memories using BF16 is valuable. If you are running out of time, I am glad to help make some changes according to Tim's suggestions based on your branch, to see whether this PR can be merged ASAP. Thank you.

Thanks, of course, you can continue to do it according to Tim's suggestions.

File: transformer_engine/pytorch/optimizers/fused_adam.py [hunk 1]
```diff
@@ -284,8 +284,11 @@ def get_unscaled_state(self, param, state_name):
             else:
                 assert state[state_name].dtype == torch.float32
             unscaled = state[state_name]
+        elif dtype == torch.bfloat16:
+            assert state[state_name].dtype == torch.bfloat16
+            unscaled = state[state_name]
```
Comment (kunlunl) [inline transformer_engine/pytorch/optimizers/fused_adam.py:289]: MCore's checkpoint needs all stetes be fp32, so I think there should cast the bf16 states to fp32?

[the remark this answers is not in the document]
Comment (young-mu) [reply transformer_engine/pytorch/optimizers/fused_adam.py:289]: BF16 has the same dynamic range as FP32 (8-bit exponent), so BF16 optimizer states do NOT need to be scaled.
In Megatron-LM's `MixedPrecisionOptimizers`, grad scaler is only applied to FP16 gradients, not BF16 gradients.

File: transformer_engine/pytorch/csrc/extensions/multi_tensor/multi_tensor_adam.cu [hunk 2]
```diff
@@ -622,22 +627,32 @@ void multi_tensor_adam_cuda(int chunk_size, at::Tensor noop_flag,
...
               g_in_type, 1, "adam",
-              multi_tensor_apply<4>((int64_t)BLOCK_SIZE, (int64_t)chunk_size, noop_flag,
-                                    tensor_lists,
-                                    AdamFunctor<scalar_t_0, scalar_t_1, float, int64_t>(), beta1,
-                                    beta2, bias_correction1, bias_correction2, epsilon, lr,
-                                    (adamMode_t)mode, weight_decay);));
+              DISPATCH_DOUBLE_FLOAT_HALF_AND_BFLOAT(
+                  m_in_type, 2, "adam",
+                  DISPATCH_DOUBLE_FLOAT_HALF_AND_BFLOAT(
+                      v_in_type, 3, "adam",
```
[the remark this answers is not in the document]
Comment (young-mu) [reply transformer_engine/pytorch/csrc/extensions/multi_tensor/multi_tensor_adam.cu:633]: 1. Yes, increasing the number of template instantiations will affect compile time and dynamic library size, but it should not impact execution time.
2. Casting the states to FP32 would indeed reduce the number of template instantiations, but it could significantly increase the optimizer step time.

File: transformer_engine/pytorch/optimizers/fused_adam.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (timmoon10) [reply transformer_engine/pytorch/optimizers/fused_adam.py:289]: I agree this behavior is optimal, but the API is very intricate and unintuitive. The user thinks they are just changing the dtype, but actually they're changing the entire algorithm. It would be a better design to have separate options for the dtype and the algorithm.

File: transformer_engine/pytorch/csrc/extensions/multi_tensor/multi_tensor_adam.cu (hunk 2 above)
[the remark this answers is not in the document]
Comment (BestJuly) [reply transformer_engine/pytorch/csrc/extensions/multi_tensor/multi_tensor_adam.cu:633]: Hi @timmoon10 , is it acceptable that we have a PR based on this one which uses bf16 optimizer while maintain the kernel in fp32?

**Comment by timmoon10** [comment]
Closed with https://github.com/NVIDIA/TransformerEngine/pull/1640.
